### ISSUE TITLE
Overridden find() and hasMany marshalling

### ISSUE THREAD

**Issue by stefanozoffoli** [cakephp/cakephp#5922]
Hi,
I'm using this SoftDeleteTrait: https://github.com/PGBI/cakephp3-soft-delete/blob/master/src/Model/Table/SoftDeleteTrait.php

When i need to patch an entity with hasmany association, Cake gives me an error here (
https://github.com/cakephp/cakephp/blob/3.0/src/ORM/Marshaller.php#L477 ) because `$indexed[$key]` is not defined anymore.

I think that when the Marshaller calculate `$maybeExistentQuery`, SoftDeleteTrait adds a condition to `$this->_table->find()`, so that `count($maybeExistentQuery->clause('where'))` returns 1 instead of 0.

Then, when Cake iterates over `$maybeExistentQuery`, it finds entities already merged and unsetted here ( https://github.com/cakephp/cakephp/blob/3.0/src/ORM/Marshaller.php#L460 )

I tried to change the code to:
```php
        if (count($maybeExistentQuery->clause('where'))) {
            foreach ($maybeExistentQuery as $entity) {
                $key = implode(';', $entity->extract($primary));
                if (isset($indexed[$key])) {
                    $output[] = $this->merge($entity, $indexed[$key], $options);
                    unset($indexed[$key]);
                }
            }
        }
```

And it works, but I don't think this is an optimal solution because it makes an unnecessary find() query in the database.

**Comment by markstory** [comment]
Adding an `isset()` seems reasonable to me. How is that a bad thing?

**Comment by stefanozoffoli** [comment]
Mmm..
I didn't understand the meaning of `$maybeExistentQuery`, but by entering the iteration because of the if statement returning true, it makes an unnecessary select to the database.

```php
if (count($maybeExistentQuery->clause('where'))) {
            foreach ($maybeExistentQuery as $entity) {
```

Result in this query:
```php
SELECT SaleDetails.id AS `SaleDetails__id`, ... SaleDetails.deleted AS `SaleDetails__deleted` FROM sale_details SaleDetails WHERE SaleDetails.deleted IS NULL
```

That results potentially in a lot of entities.

**Comment by markstory** [comment]
OK, I see the issue you're getting at now. The original find query wouldn't have had conditions, but the overloaded find() adds in a condition that selects far too many rows. In your situation is `$indexed` empty?

**Comment by stefanozoffoli** [comment]
Yes `$indexed` becomes empty right before `$maybeExistentQuery` definition!

**Comment by markstory** [comment]
Cool, I can put together a fix then.

**Comment by stefanozoffoli** [comment]
Thank you :smile:

**Comment by markstory** [comment]
Fixed with #5928
